plugins/mana/plugin.go

```diff
@@ -371,6 +395,91 @@ func GetPendingMana(value float64, n time.Duration) float64 {
 	return value * (1 - math.Pow(math.E, -mana.Decay*(n.Seconds())))
 }
 
+// GetPastConsensusManaVector builds a consensus base mana vector in the past.
+func GetPastConsensusManaVector(t time.Time) (*mana.ConsensusBaseManaVector, error) {
+	cbmvPast := &mana.ConsensusBaseManaVector{}
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+
+	if cachedMetadata.Exists() {
+		metadata := cachedMetadata.Unwrap()
+		if t.After(metadata.Timestamp) {
+			consensusBaseManaPastVectorStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+				cachedPbm := &mana.CachedPersistableBaseMana{CachedObject: cachedObject}
+				defer cachedPbm.Release()
+				p := cachedPbm.Unwrap()
+				err := cbmvPast.FromPersistable(p)
+				if err != nil {
+					log.Errorf("error while restoring %s mana vector from storage: %w", mana.ConsensusMana.String(), err)
+					cbmvPast = &mana.ConsensusBaseManaVector{}
+					return false
+				}
+				return true
+			})
+		}
+	}
+
+	var eventLogs []mana.Event
+	var readErr error
+	consensusEventsLogStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
+		pe := cachedPe.Unwrap()
+		defer cachedPe.Release()
+		if pe.Time.After(t) {
+			return false
+		}
+
+		// already consumed in stored base mana vector.
+		if cachedMetadata.Exists() && cbmvPast.Size() > 0 {
+			metadata := cachedMetadata.Unwrap()
+			if pe.Time.Before(metadata.Timestamp) {
+				return true
+			}
+		}
+
+		ev, err := mana.FromPersistableEvent(pe)
+		if err != nil {
+			readErr = err
+			return false
+		}
+		eventLogs = append(eventLogs, ev)
+		return true
+	})
+	if readErr != nil {
+		return nil, readErr
+	}
+	index, err := cbmvPast.BuildPastBaseVector(eventLogs, t)
+	if err != nil {
+		return nil, err
+	}
+
+	// store the most recent bmv
+	if err = consensusBaseManaPastVectorStorage.Prune(); err != nil {
+		return nil, err
+	}
+	for _, p := range cbmvPast.ToPersistables() {
+		consensusBaseManaPastVectorStorage.Store(p).Release()
+	}
+
+	//store the metadata
+	metadata := &mana.ConsensusBasePastManaVectorMetadata{
+		Timestamp: t,
+		Index:     int64(index),
+	}
+
+	if !cachedMetadata.Exists() {
+		consensusBaseManaPastVectorMetadataStorage.Store(metadata)
+	} else {
+		m := cachedMetadata.Unwrap()
+		m.Update(metadata)
+	}
+
+	// TODO: delete logs? what if we want to go further in the past
```

Comment:
Logs should be deleted after some time from the object storage, but let's archive them into something standalone, like a csv file or mongoDB or whatever. So even if the node crashes, db is dirty and deleted, we have them somewhere for analysis.